tests/utils_tests/test_html.py

```diff
@@ -182,3 +181,55 @@ def test_conditional_escape(self):
         self.assertEqual(html.conditional_escape(s),
                          '&lt;h1&gt;interop&lt;/h1&gt;')
         self.assertEqual(html.conditional_escape(safestring.mark_safe(s)), s)
+
+    def test_html_safe(self):
+        @html.html_safe
+        class HtmlClassOne(object):
+            def __str__(self):
+                return "<h1>I'm html class one!</h1>"
+
+        @html.html_safe
+        class HtmlClassTwo(object):
+            def __unicode__(self):
+                return "<h1>I'm html class two!</h1>"
+
+        html_one = HtmlClassOne()
+        html_two = HtmlClassTwo()
+        self.assertTrue(hasattr(HtmlClassOne, '__html__'))
+        self.assertTrue(hasattr(html_one, '__html__'))
+        self.assertEqual(force_text(html_one), html_one.__html__())
+        self.assertTrue(hasattr(HtmlClassTwo, '__html__'))
+        self.assertTrue(hasattr(html_two, '__html__'))
+        self.assertEqual(force_text(html_two), html_two.__html__())
+
+    def test_html_safe_subclass(self):
+        class BaseClass(object):
+            def __html__(self):
+                # defines __html__ on its own
+                return 'some html content'
+
+            def __str__(self):
+                return 'some non html content'
+
+        @html.html_safe
+        class Subclass(BaseClass):
+            def __str__(self):
+                # overrides __str__ and is marked as html_safe
+                return 'some html safe content'
+
+        subclass_obj = Subclass()
+        self.assertEqual(force_text(subclass_obj), subclass_obj.__html__())
+
+    def test_html_safe_errors(self):
+        with self.assertRaisesMessage(ValueError, "can't apply @html_safe to "
+                "HtmlClassOne because it defines __html__()."):
+            @html.html_safe
+            class HtmlClassOne(object):
+                def __html__(self):
+                    return "<h1>I'm a html class!</h1>"
+
+        with self.assertRaisesMessage(ValueError, "can't apply @html_safe to "
```

Comment:
For better readability rather than splitting the string across two lines:
```
msg = "can't apply.. "
with self.assertRaisesMessage(ValueError, msg):
```